### PR TITLE
Offer safe PLL's

### PR DESCRIPTION
**Pull request by DigitalBrains1**
We have the convention that if a function can lead to timing violations, we prefix the name of that function with `unsafe`. The functions in `Clash.Intel.ClockGen` and `Clash.Xilinx.ClockGen` before this PR do not follow that convention. The `locked` output of these clock generators is an asynchronous signal. All our examples showed the correct way to handle this, but it should be called out by an `unsafe` prefix, and we should offer an easy-to-use alternative. It was too easy for a user to forget to synchronise the `locked` signal.

This PR deprecates the Intel functions and adds new functions, both a safe variant and, for advanced use cases, an unsafe variant. The safe variant incorporates `resetSynchronizer` so there's just a ready-to-use `Reset` instead of the asynchronous `locked` signal.

For Xilinx, it was decided to not deprecate the functions but rather scrap them altogether and provide new, safe ones with the same name as the old unsafe ones. The functions in Clash 1.6 are so broken that keeping them around as deprecated functions doesn't add much utility for our users; better to bite the bullet. PR #2427 had already changed the API for Xilinx, by the way.

* All documentation and examples now shows the safe variants where that makes sense. Since we recommend people always use a PLL, in practice that meant we recommended people to use functions like `unsafeFromActiveLow`. Now people can follow the recommendation without being exposed to `unsafe` stuff.
* All clock generators support multiple output clocks.
* It is no longer required to give a clock generator a name. Instead, people can set the name of the instance with `Clash.Magic.setName` if they need a predictable name.
* Intel PLL's now account for the `ResetPolarity` of the input domain instead of always expecting an active-high signal.

## Still TODO:

  - [x] Write a changelog entry (see changelog/README.md)
  - [x] Check copyright notices are up to date in edited files

### CODE REVIEW

File: clash-prelude/src/Clash/Intel/ClockGen.hs [hunk 1]
```diff
@@ -49,78 +68,255 @@ import Clash.Promoted.Symbol (SSymbol)
...
+If you haven't determined the frequency you want the design to run at, the
+predefined 100 MHz domain t'Clash.Signal.XilinxSystem' can be a good starting
+point. The datasheet for your FPGA specifies lower and upper limits, but the
+true maximum frequency is determined by your design.
+
+Supposing you need a clock running at 150 MHz for your design, the following
+will instantiate a suitable domain:
+
+@
+'Clash.Signal.createDomain' 'Clash.Signal.vXilinxSystem'{vName=\"Dom150\", vPeriod='Clash.Signal.hzToPeriod' 150e6}
```
Comment (christiaanb) [inline clash-prelude/src/Clash/Intel/ClockGen.hs:98]: Perhaps we should replace `Xilinx` by `Intel` in this module.

File: clash-prelude/src/Clash/Intel/ClockGen.hs [hunk 2]
```diff
@@ -49,78 +68,255 @@ import Clash.Promoted.Symbol (SSymbol)
...
+@
+
+@Dom150@ will have 'Clash.Signal.Synchronous' resets on its memory elements
```
Comment (christiaanb) [inline clash-prelude/src/Clash/Intel/ClockGen.hs:101]: Perhaps leave this part out, only from Stratix 10 onwards does Intel recommend synchronous resets.

File: clash-prelude/src/Clash/Intel/ClockGen.hs [hunk 3]
```diff
@@ -49,78 +68,255 @@ import Clash.Promoted.Symbol (SSymbol)
...
+generator will pick the attainable frequency closest to the requested frequency
+(or possibly fail to synthesize). You can check the frequency that the wizard
+chose by loading your design into the Vivado GUI. In the /IP sources/ window,
+choose the clock wizard and select /Re-customize IP.../. On the /Output Clocks/
+tab, the relevant column is /Actual Output Freq (MHz)/. If the actual value
+differs, copy the actual value back to the Clash design.
+-}
+
+{- $using
+The functions in this module will instantiate a Xilinx MMCM clock generator
```
Comment (christiaanb) [inline clash-prelude/src/Clash/Intel/ClockGen.hs:122]: This part doesn't really apply to `Intel`

**Review by christiaanb — APPROVED** [review]
LGTM

Only some minor details w.r.t. documentation: The Intel docs say things about Xilinx.

**Comment by DigitalBrains1** [comment]
Ah, I had neglected to mention that I was still working on the Intel documentation, I simply copied the Xilinx one and still need to adjust those things.

**Comment by DigitalBrains1** [comment]
Since your review, I've made the following changes:
* Finished documentation for `Clash.Intel.ClockGen`
* For both Xilinx and Intel: switched the example to use the implicit `Prelude` as I think this is more familiar to a novice
* Again for both: added an example of the use of the `unsafe` functions
* Adjusted the commit message for the _Improve readability of Clash.Clocks.Deriving_ commit, from _No functional change_ to:
    > No functional change, except that the `clocks` function is now annotated with a `CLASH_OPAQUE` "pragma" rather than a mere `NOINLINE`.

    We discovered this had been overlooked for PR #2511, but since this PR fundamentally changed the code, we decided to adjust it here.

File: clash-prelude/src/Clash/Intel/ClockGen.hs [hunk 4]
```diff
@@ -49,78 +70,275 @@ import Clash.Promoted.Symbol (SSymbol)
...
+chose by loading your design into the Quartus GUI. In the /Project Navigator/,
+choose the /Hierarchy/ view and find your clock generator instance.
+Double-click the instance to open Platform Designer and choose /Edit/
+/Parameters.../. In the /Output Clocks/ page, the relevant column is /Actual/
+/Settings/. If the actual value differs, copy the actual value back to the
+Clash design.
+-}
+
+{- $using
+The functions in this module will instantiate an Intel IP core with 1 reference
```
[the remark this answers is not in the document]
Comment (DigitalBrains1) [reply clash-prelude/src/Clash/Intel/ClockGen.hs:123]: It is intended as a generic comment. Intel calls these things "IP cores" (and luckily so does Xilinx, wonderful when it aligns). We instantiate the IP core for a PLL.

[the remark this answers is not in the document]
Comment (DigitalBrains1) [reply clash-prelude/src/Clash/Intel/ClockGen.hs:123]: When you look at the individual functions, it'll say
> Instantiate an Intel clock generator corresponding to the Intel/Quartus "ALTPLL" IP core

> Instantiate an Intel clock generator corresponding to the Intel/Quartus "Altera PLL" IP core

summing those two up into a single sentence led me to writing
> The functions in this module will instantiate an Intel IP core

File: clash-prelude/src/Clash/Intel/ClockGen.hs [hunk 5]
```diff
@@ -5,36 +5,57 @@ Copyright  :  (C) 2017-2018, Google Inc
 License    :  BSD2 (see the file LICENSE)
 Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
 
-PLL and other clock-related components for Intel (Altera) FPGAs
+This module contains functions for instantiating clock generators on Intel
+FPGA's.
 
-A PLL generates a stable clock signal for your circuit at a selectable
-frequency.
+We suggest you always use a clock generator even if your oscillator runs at the
```
[the remark this answers is not in the document]
Comment (DigitalBrains1) [reply clash-prelude/src/Clash/Intel/ClockGen.hs]: Agreed

**Pull request merged**